### DISCUSSION TITLE
How to read stdin (and why in that way)

### DISCUSSION THREAD

**Discussion by Devilly** [pybricks/support#1108] [Q&A] [answered]
On the page about [keyboard to hub communication](https://pybricks.com/projects/tutorials/wireless/hub-to-device/pc-keyboard/) there is the use of `if keyboard.poll(0):` It looks to me, based on testing, the functionality works just as well without the polling. Therefore the question, what does the polling contribute to?

And then there is the [device to hub communication](https://pybricks.com/projects/tutorials/wireless/hub-to-device/pc-communication/), where not `stdin.read` is used (just like in the first link) but `stdin.buffer.read`, which I can not explain. I tried looking for the difference in the API docs but couldn't find any relevant information at all. Can someone explain to me how these compare or tell me where the documentation can be found.

And then, last of all: I just started with Pybricks and I'm really liking it so far. The development experience is really nice, thanks a lot to the devs for making this a possibility!

**Comment by dlech** [comment, marked as answer]
The example is minimal, so doesn't really show the usefulness. The point is to avoid blocking reads so that you can still run a control loop to control motors with feedback from sensors any only read when data is actually available.

`stdin.read()` returns a `str` type object while `stdin.buffer.read()` returns a `bytes` object. If your data is not valid UTF-8 encoded text (e.g. arbitrary binary data) then you should use the latter.

In our next release, we are introducing new ways to do async code and a simpler way to read one byte from stdin (https://github.com/pybricks/pybricks-micropython/pull/177) so we probably be updating all of the examples to no longer use `poll()`.

**Reply by Devilly** [reply]
Very clear, many thanks. :)

**Comment by vascolp** [comment]
Just noticed that the [device to hub communication](https://pybricks.com/projects/tutorials/wireless/hub-to-device/pc-communication/) does not work with 3.3.0b6 but it works with 3.2.3.

Is this expected, will there be a new way to comunicate?

**Reply by dlech** [reply]
Yes, it is expected since we made some changes in the communication profile. 

Here is an updated example for the v3.3 firmware:

```python
# SPDX-License-Identifier: MIT
# Copyright (c) 2020 Henrik Blidh
# Copyright (c) 2022-2023 The Pybricks Authors

"""
Example program for computer-to-hub communication.

Requires Pybricks firmware >= 3.3.0b4.
"""

import asyncio
from contextlib import suppress
from bleak import BleakScanner, BleakClient

PYBRICKS_COMMAND_EVENT_CHAR_UUID = "c5f50002-8280-46da-89f4-6d8051e4aeef"

# Replace this with the name of your hub if you changed
# it when installing the Pybricks firmware.
HUB_NAME = "Pybricks Hub"


async def main():
    main_task = asyncio.current_task()

    def handle_disconnect(_):
        print("Hub was disconnected.")

        # If the hub disconnects before this program is done,
        # cancel this program so it doesn't get stuck waiting
        # forever.
        if not main_task.done():
            main_task.cancel()

    ready_event = asyncio.Event()

    def handle_rx(_, data: bytearray):
        if data[0] == 0x01:  # "write stdout" event (0x01)
            payload = data[1:]

            if payload == b"rdy":
                ready_event.set()
            else:
                print("Received:", payload)

    # Do a Bluetooth scan to find the hub.
    device = await BleakScanner.find_device_by_name(HUB_NAME)

    if device is None:
        print(f"could not find hub with name: {HUB_NAME}")
        return

    # Connect to the hub.
    async with BleakClient(device, handle_disconnect) as client:

        # Subscribe to notifications from the hub.
        await client.start_notify(PYBRICKS_COMMAND_EVENT_CHAR_UUID, handle_rx)

        # Shorthand for sending some data to the hub.
        async def send(data):
            # Wait for hub to say that it is ready to receive data.
            await ready_event.wait()
            # Prepare for the next ready event.
            ready_event.clear()
            # Send the data to the hub.
            await client.write_gatt_char(
                PYBRICKS_COMMAND_EVENT_CHAR_UUID,
                b"\x06" + data,  # prepend "write stdin" command (0x06)
                response=True
            )

        # Tell user to start program on the hub.
        print("Start the program on the hub now with the button.")

        # Send a few messages to the hub.
        for i in range(5):
            await send(b"fwd")
            await asyncio.sleep(1)
            await send(b"rev")
            await asyncio.sleep(1)
            print(".", end="", flush=True)

        # Send a message to indicate stop.
        await send(b"bye")

        print("done.")

    # Hub disconnects here when async with block exits.


# Run the main async program.
if __name__ == "__main__":
    with suppress(asyncio.CancelledError):
        asyncio.run(main())
```

**Comment by vascolp** [comment]
Good, thanks for the code!
That is the pc side code. I tried with the hub side code that I got from the project demos and could not get it to work... 

Also BleakScanner.find_device_by_name is very recent, I had to upgrade my bleak.

**Reply by dlech** [reply]
Here is the updated hub side program:

```python
# NOTE: Run this program with the latest
# firmware provided via https://beta.pybricks.com/

from pybricks.pupdevices import Motor
from pybricks.parameters import Port
from pybricks.tools import wait

# Standard MicroPython modules
from usys import stdin, stdout
from uselect import poll

motor = Motor(Port.A)

# Optional: Register stdin for polling. This allows
# you to wait for incoming data without blocking.
keyboard = poll()
keyboard.register(stdin)

while True:

    # Let the remote program know we are ready for a command.
    stdout.buffer.write(b"rdy")

    # Optional: Check available input.
    while not keyboard.poll(0):
        # Optional: Do something here.
        wait(10)

    # Read three bytes.
    cmd = stdin.buffer.read(3)

    # Decide what to do based on the command.
    if cmd == b"fwd":
        motor.dc(50)
    elif cmd == b"rev":
        motor.dc(-50)
    elif cmd == b"bye":
        break
    else:
        motor.stop()
```

**Comment by vascolp** [comment]
Yes I have been playing a bit more and reach that conclusion!
rdy in the beggining instead of OK in the end.
Thnak you for you fast reply!

**Comment by vascolp** [comment]
Sorry to wake this thread again but...

Instead of using this:
```
       # Tell user to start program on the hub.
        print("Start the program on the hub now with the button.")
```
is there a way to start the program execution via BLE? 
I tried this but it did not work...
```
        await self.client.write_gatt_char(PYBRICKS_COMMAND_EVENT_CHAR_UUID,
            struct.pack("<B", 1), #Command.START_USER_PROGRAM),
                                          response=False)
```
I am sure there is a solution for this, pybricks.code does it all the time...
Thanks!!!!

**Reply by dlech** [reply]
It should be `response=True`. And it assumes there is already a valid program loaded in RAM.

**Comment by vascolp** [comment]
My bug... (no class, self. shouldn't be there... copy paste problems)
Thank you and sorry to bother!

**Comment by BY571** [comment]
@dlech Can you elaborate a bit on how I would need to adapt your example when I would want to send and receive float values to my hub?

This would be example client.py script
```python
from pybricks.hubs import InventorHub
from pybricks.pupdevices import Motor, UltrasonicSensor,
from pybricks.parameters import Port
from pybricks.tools import wait

# Standard MicroPython modules
from usys import stdin, stdout
from uselect import poll
import ustruct
from micropython import kbd_intr

kbd_intr(-1)

hub = InventorHub()

motor = Motor(Port.A)
eyes = UltrasonicSensor(Port.E)

keyboard = poll()
keyboard.register(stdin)


while True:
    
    while not keyboard.poll(0):
        wait(1)

    data = stdin.buffer.read(4)
    value = ustruct.unpack("!f", data)[0]
    motor.run_target(200, target_angle=179*value)


    motor_angle = motor.angle())

    (pitch, roll) = hub.imu.tilt()

    dist = eyes.distance()

    # send current state
    out_msg = ustruct.pack('!ffff', motor_angle, pitch, roll, dist)
    stdout.buffer.write(out_msg)
```

my send function seems to work but I think I have issues reading the data:

```python
    async def _send(self, data: bytes) -> None:
        try:
            # Send some data to the hub.
            await self.client.write_gatt_char(
                self.PYBRICKS_COMMAND_EVENT_CHAR_UUID,
                b"\x06" + data,  # Prepend "write stdin" command 
                response=True,
            )
        except Exception as e:
            # Handle exceptions.
            print(e)
            await self.disconnect()
            
     async def _handle_rx(self, _, data: bytes) -> None:
        # add received data to the queue
        if data[0] == 0x01:  # "write stdout" event (0x01)
            payload = data[1:]
            print("Received:", payload)
            data = payload
            await self.rx_queue.put(payload)
            
```

**Reply by dlech** [reply]
You would just use `struct.pack()` and `struct.unpack()` on the PC side with the same format args.

**Reply by BY571** [reply]
I think I found the issue. It seems that the data (7 float values) exceeds this MTU size. I could read it but only in two chunks and not in one. So I probably have to switch to integer values.
